src/sources/prometheus/mod.rs

```diff
@@ -0,0 +1,210 @@
+use crate::{topology::config::GlobalOptions, Event};
+use futures::{sync::mpsc, Future, Sink, Stream};
+use http::Uri;
+use hyper;
+use hyper_tls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::time::{Duration, Instant};
+use tokio::timer::Interval;
+
+pub mod parser;
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct PrometheusConfig {
+    hosts: Vec<String>,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+#[typetag::serde(name = "prometheus")]
+impl crate::topology::config::SourceConfig for PrometheusConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let mut urls = Vec::new();
+        for host in self.hosts.iter() {
+            let base_uri = host.parse::<Uri>().context(super::UriParseError)?;
+            urls.push(format!("{}metrics", base_uri));
+        }
+        Ok(prometheus(urls, self.scrape_interval_secs, out))
+    }
+
+    fn output_type(&self) -> crate::topology::config::DataType {
+        crate::topology::config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "prometheus"
+    }
+}
+
+fn prometheus(urls: Vec<String>, interval: u64, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending metric: {:?}", e));
+
+    let task = Interval::new(Instant::now(), Duration::from_secs(interval))
+        .map_err(|e| error!("timer error: {:?}", e))
+        .map(move |_| futures::stream::iter_ok(urls.clone()))
+        .flatten()
+        .map(move |url| {
+            let https = HttpsConnector::new(4).expect("TLS initialization failed");
+            let client = hyper::Client::builder().build(https);
+
+            let request = hyper::Request::get(&url)
+                .body(hyper::Body::empty())
+                .unwrap();
```

Comment:
If we never expect this to fail it'd be nice to use `expect` with a message, just in case.